crates/ruma-macros/src/lib.rs

```diff
@@ -108,27 +215,194 @@ pub fn event_enum(input: TokenStream) -> TokenStream {
+///
+/// Some kinds can generate a modified clone of the event content type. For instance, for an event
+/// content type named `FooEventContent`:
+///
+/// * `RedactedFooEventContent`: the redacted form of the event content, for the `MessageLike` and
+///   `StateKind`. It also generates the `RedactContent` implementation which applies the redaction
+///   algorithm according to the Matrix specification.
+///
+///   The generated type implements `Redacted{Kind}EventContent`, `StaticEventContent`, `Serialize`
+///   and `Deserialize`.
+///
+///   The generation only works if the type is a struct with named fields. To keep a field after
+///   redaction, the `#[ruma_event(skip_redaction)]` attribute can be applied to that field.
+///
+///   To skip the generation of this type and trait to implement a custom redaction, or because it
+///   is not a struct with named fields, the `#[ruma_event(custom_redacted)]` attribute can be used
+///   on the container. The `RedactedFooEventContent` type must still exist and implement the same
+///   traits, even if it is only a type alias, and the `RedactContent` trait must still be
+///   implemented for those kinds.
+/// * `PossiblyRedactedFooEventContent`: the form of the event content that is used when we don't
+///   know whether a `State` event is redacted or not. It means that on this type any field that is
+///   redacted must be optional, or it must have the `#[serde(default)]` attribute for
+///   deserialization.
 ///
-/// Also generates type aliases depending on the kind of event, with the final `Content` of the type
-/// name removed and prefixed added. For instance, a message-like event content type
-/// `FooEventContent` will have the following aliases generated:
+///   The generated type implements `PossiblyRedactedStateEventContent`, `StaticEventContent`,
+///   `Serialize` and `Deserialize`.
 ///
-/// * `type FooEvent = MessageLikeEvent<FooEventContent>`
-/// * `type SyncFooEvent = SyncMessageLikeEvent<FooEventContent>`
-/// * `type OriginalFooEvent = OriginalMessageLikeEvent<FooEventContent>`
-/// * `type OriginalSyncFooEvent = OriginalSyncMessageLikeEvent<FooEventContent>`
-/// * `type RedactedFooEvent = RedactedMessageLikeEvent<FooEventContent>`
-/// * `type RedactedSyncFooEvent = RedactedSyncMessageLikeEvent<FooEventContent>`
+///   The generation uses the rules as the redacted type, using the `#[ruma_event(skip_redaction)]`
+///   attribute.
+///
+///   To skip the generation of this type to use a custom type, the
+///   `#[ruma_event(custom_possibly_redacted)]` attribute can be used on the container. The
+///   `PossiblyRedactedFooEventContent` type must still exist for the `State` kind and implement the
+///   same traits, even if it is only a type alias.
+///
+/// Event content types of the `MessageLike` kind that use the `Relation` type also need a clone of
+/// the event content without the `relates_to` field. This macro can generate a
```

Comment:
Maybe note what this is useful for? "for use within relations, where nested relations are not meant to be serialized by homeservers" or something like that.

Reply:
Applied.